src/main/java/driver/DriverFactory.java

```diff
@@ -239,4 +280,69 @@ public void openAddLicensesPage() {
         addLicenseButton.click();
         pageToLoad(Constants.LICENSE_DETAILS_URL);
     }
-}
\ No newline at end of file
+    public String getTextFromTable(){
+        return driver.findElement(By.id("device-table")).getText();
+    }
+
+    public boolean hasRequired(WebElement element){
+        return  Boolean.parseBoolean(element.getAttribute("required"));
+    }
+    public void fillAllFieldsDevice(){
+        deviceName = randomName();
+        serialNo = randomNumber();
+        inventoryNo = randomNumber();
+        invoiceNo = randomNumber();
+        addDeviceButton.click();
+        nameField.sendKeys(deviceName);
+        serialNoField.sendKeys(serialNo + "");
+        inventoryNoField.sendKeys(inventoryNo + "");
+        invoiceNoField.sendKeys(invoiceNo + "");
+        descriptionField.sendKeys(Constants.DEVICE_DESCRIPTION);
+        subcategoryList.click();
+        subcategoryLaptopListElement.click();
+        tagsList.click();
+        tagsListElement.click();
+        locationsList.click();
+        locationsRomaniaListElement.click();
+        saveDeviceButton.click();
+    }
+    public void fillAllMandatoryFields(WebElement field){
+        deviceName = randomName();
+        serialNo = randomNumber();
+        inventoryNo = randomNumber();
+        invoiceNo = randomNumber();
+        addDeviceButton.click();
+        nameField.sendKeys(deviceName);
+        serialNoField.sendKeys(serialNo + "");
+        inventoryNoField.sendKeys(inventoryNo + "");
+        invoiceNoField.sendKeys(invoiceNo + "");
+        clearField(field);
+    }
+
+
+    public void openColumns(WebElement element) {
+        columnsButton.click();
+        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", element);
+        elementToLoad(element);
+        element.click();
+        WebElement closeColumns = driver.switchTo().activeElement();
+        closeColumns.sendKeys(Keys.ESCAPE);
+    }
+
+    public void openThreeDotsOptions(String dropdown, String option){
+        dropdown = deviceDropdown + deviceName;
+        elementToLoadForComposedElements(dropdown);
+        driver.findElement(By.id(dropdown)).click();
+        option = option + deviceName;
+        elementToLoadForComposedElements(option);
+        driver.findElement(By.id(option)).click();
+    }
+    public void openThreeDotsOptionsForInactive(String dropdown, String inactive){
+        dropdown = deviceDropdown + inactiveDeviceName;
+        elementToLoadForComposedElements(dropdown);
+        driver.findElement(By.id(dropdown)).click();
+        inactive = assignDevice + inactiveDeviceName;
+        elementToLoadForComposedElements(inactive);
+        driver.findElement(By.id(inactive)).click();
+    }
```

Comment:
i dont think these work as you expect them to. for example: openThreeDotsOptions(deviceDropdown,editDevice) results in "device_dropdown_" = "device_dropdown_" + deviceName